File: avl-tree-list/avltreelist-test.rs
/* 
 * AVL tree list test (Rust)
 * 
 * Copyright (c) 2019 Project Nayuki. (MIT License)
 * https://www.nayuki.io/page/avl-tree-list
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * - The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 * - The Software is provided "as is", without warranty of any kind, express or
 *   implied, including but not limited to the warranties of merchantability,
 *   fitness for a particular purpose and noninfringement. In no event shall the
 *   authors or copyright holders be liable for any claim, damages or other
 *   liability, whether in an action of contract, tort or otherwise, arising from,
 *   out of or in connection with the Software or the use or other dealings in the
 *   Software.
 */

extern crate rand;
use rand::distributions::IndependentSample;
use rand::distributions::range::Range;
mod avltreelist;
use avltreelist::AvlTreeList;


fn main() {
	test_add();
	test_add_list();
	test_set();
	test_insert_at_beginning();
	test_insert_at_end();
	test_insert_at_middle();
	test_insert_list();
	test_insert_many_beginning();
	test_insert_many_end();
	test_insert_many_everywhere();
	test_remove();
	test_clear();
	test_iterator();
	test_against_rust_vec_randomly();
	println!("Test passed");
}


fn test_add() {
	let mut list = AvlTreeList::<&'static str>::new();
	list.push("January");
	list.push("February");
	list.push("March");
	list.push("April");
	list.push("May");
	list.push("June");
	list.check_structure();
	assert_eq!(list.len(), 6);
	assert_eq!(list[0], "January" );
	assert_eq!(list[1], "February");
	assert_eq!(list[2], "March"   );
	assert_eq!(list[3], "April"   );
	assert_eq!(list[4], "May"     );
	assert_eq!(list[5], "June"    );
}


fn test_add_list() {
	let mut list = AvlTreeList::<&'static str>::new();
	list.append(&mut vec!["January"]);
	list.append(&mut vec!["February", "March", "April"]);
	list.append(&mut vec!["May", "June", "July", "August", "September", "October", "November", "December"]);
	assert_eq!(list.len(), 12);
	assert_eq!(list[ 0], "January"  );
	assert_eq!(list[ 1], "February" );
	assert_eq!(list[ 2], "March"    );
	assert_eq!(list[ 3], "April"    );
	assert_eq!(list[ 4], "May"      );
	assert_eq!(list[ 5], "June"     );
	assert_eq!(list[ 6], "July"     );
	assert_eq!(list[ 7], "August"   );
	assert_eq!(list[ 8], "September");
	assert_eq!(list[ 9], "October"  );
	assert_eq!(list[10], "November" );
	assert_eq!(list[11], "December" );
}


fn test_set() {
	let mut list = AvlTreeList::<&'static str>::new();
	for _ in 0 .. 10 {
		list.push("");
	}
	list[0] = "zero";
	list[1] = "ten";
	list[2] = "twenty";
	list[3] = "thirty";
	list[4] = "forty";
	list[5] = "fifty";
	list[6] = "sixty";
	list[7] = "seventy";
	list[8] = "eighty";
	list[9] = "ninety";
	assert_eq!(list.len(), 10);
	assert_eq!(list[0], "zero"   );
	assert_eq!(list[1], "ten"    );
	assert_eq!(list[2], "twenty" );
	assert_eq!(list[3], "thirty" );
	assert_eq!(list[4], "forty"  );
	assert_eq!(list[5], "fifty"  );
	assert_eq!(list[6], "sixty"  );
	assert_eq!(list[7], "seventy");
	assert_eq!(list[8], "eighty" );
	assert_eq!(list[9], "ninety" );
}


fn test_insert_at_beginning() {
	let mut list = AvlTreeList::<&'static str>::new();
	list.insert(0, "Sunday");
	list.insert(0, "Monday");
	list.insert(0, "Tuesday");
	assert_eq!(list.len(), 3);
	assert_eq!(list[0], "Tuesday");
	assert_eq!(list[1], "Monday" );
	assert_eq!(list[2], "Sunday" );
}


fn test_insert_at_end() {
	let mut list = AvlTreeList::<&'static str>::new();
	list.insert(0, "Saturday");
	list.insert(1, "Friday");
	list.insert(2, "Thursday");
	list.insert(3, "Wednesday");
	assert_eq!(list.len(), 4);
	assert_eq!(list[0], "Saturday" );
	assert_eq!(list[1], "Friday"   );
	assert_eq!(list[2], "Thursday" );
	assert_eq!(list[3], "Wednesday");
}


fn test_insert_at_middle() {
	let mut list = AvlTreeList::<&'static str>::new();
	list.insert(0, "Up");
	list.insert(1, "Down");
	list.insert(1, "Left");
	list.insert(2, "Right");
	list.insert(1, "Front");
	list.insert(2, "Back");
	assert_eq!(list.len(), 6);
	assert_eq!(list[0], "Up"   );
	assert_eq!(list[1], "Front");
	assert_eq!(list[2], "Back" );
	assert_eq!(list[3], "Left" );
	assert_eq!(list[4], "Right");
	assert_eq!(list[5], "Down" );
}


fn test_insert_list() {
	let mut list = AvlTreeList::<&'static str>::new();
	list.append_at(0, &mut vec!["1", "2", "3", "5"]);
	list.append_at(4, &mut vec!["377", "610", "987"]);
	list.append_at(4, &mut vec!["8", "13", "21", "144", "233"]);
	list.append_at(7, &mut vec!["34", "55", "89"]);
	assert_eq!(list.len(), 15);
	assert_eq!(list[ 0],  "1");
	assert_eq!(list[ 1],  "2");
	assert_eq!(list[ 2],  "3");
	assert_eq!(list[ 3],  "5");
	assert_eq!(list[ 4],  "8");
	assert_eq!(list[ 5], "13");
	assert_eq!(list[ 6], "21");
	assert_eq!(list[ 7], "34");
	assert_eq!(list[ 8], "55");
	assert_eq!(list[ 9], "89");
	assert_eq!(list[10], "144");
	assert_eq!(list[11], "233");
	assert_eq!(list[12], "377");
	assert_eq!(list[13], "610");
	assert_eq!(list[14], "987");
}


// Stresses the self-balancing mechanism
fn test_insert_many_beginning() {
	let n: i32 = 300000;
	let mut list = AvlTreeList::<i32>::new();
	for i in 0 .. n {
		list.push(i);
	}
	
	for (i, x) in (0i32 .. ).zip(list.into_iter()) {
		assert_eq!(*x, i);
	}
}


// Stresses the self-balancing mechanism
fn test_insert_many_end() {
	let n: i32 = 300000;
	let mut list = AvlTreeList::<i32>::new();
	for i in (0 .. n).rev() {
		list.insert(0, i);
	}
	
	for (i, x) in (0i32 .. ).zip(list.into_iter()) {
		assert_eq!(*x, i);
	}
}


// Adds in a weird binary pattern to stress arrays and linked lists
fn test_insert_many_everywhere() {
	let n: i32 = 18;
	let mut list = AvlTreeList::<i32>::new();
	list.push(0);
	for i in (0 .. n).rev() {
		let mut j: i32 = 1 << i;
		let mut k: usize = 1;
		while j < (1 << n) {
			list.insert(k, j);
			j += 2 << i;
			k += 2;
		}
	}
	
	for (i, x) in (0i32 .. ).zip(list.into_iter()) {
		assert_eq!(*x, i);
	}
}


fn test_remove() {
	let mut list = AvlTreeList::<char>::new();
	let s = "the quick brown fox jumped over the lazy dog";
	for c in s.chars() {
		list.push(c);
	}
	assert_eq!(list.len(), s.len());
	
	assert_eq!(list.remove( 2), 'e');
	assert_eq!(list.remove( 4), 'u');
	assert_eq!(list.remove( 3), 'q');
	assert_eq!(list.remove( 2), ' ');
	assert_eq!(list.remove(12), 'f');
	assert_eq!(list.remove(11), ' ');
	assert_eq!(list.remove(10), 'n');
	assert_eq!(list.remove( 9), 'w');
	assert_eq!(list.remove(11), ' ');
	assert_eq!(list.remove(11), 'j');
	assert_eq!(list.remove(11), 'u');
	assert_eq!(list.remove(10), 'x');
	assert_eq!(list.remove(11), 'p');
	assert_eq!(list.remove(12), 'd');
	assert_eq!(list.remove(11), 'e');
	assert_eq!(list.remove(13), 'v');
	assert_eq!(list.remove(13), 'e');
	assert_eq!(list.remove(19), 'l');
	assert_eq!(list.remove(20), 'z');
	assert_eq!(list.remove(19), 'a');
	assert_eq!(list.remove(18), ' ');
	assert_eq!(list.remove(22), 'g');
	
	let s = "thick broom or they do";
	assert_eq!(list.len(), s.len());
	for (i, c) in (0usize .. ).zip(s.chars()) {
		assert_eq!(list[i], c);
	}
	
	assert_eq!(list.remove(0), 't');
	assert_eq!(list.remove(2), 'c');
	assert_eq!(list.remove(2), 'k');
	assert_eq!(list.remove(2), ' ');
	assert_eq!(list.remove(2), 'b');
	assert_eq!(list.remove(2), 'r');
	assert_eq!(list.remove(2), 'o');
	assert_eq!(list.remove(2), 'o');
	assert_eq!(list.remove(4), 'o');
	assert_eq!(list.remove(7), 'h');
	assert_eq!(list.remove(5), ' ');
	assert_eq!(list.remove(5), 't');
	assert_eq!(list.remove(9), 'o');
	assert_eq!(list.remove(7), ' ');
	assert_eq!(list.remove(6), 'y');
	
	let s = "him red";
	for (i, c) in (0usize .. ).zip(s.chars()) {
		assert_eq!(list[i], c);
	}
}


fn test_clear() {
	let mut list = AvlTreeList::<i32>::new();
	for i in 0i32 .. 20 {
		list.push(i * i);
	}
	
	list.clear();
	assert_eq!(list.len(), 0);
	
	list.push(- 1);
	list.push(- 8);
	list.push(-27);
	assert_eq!(list.len(), 3);
	assert_eq!(list[0], - 1);
	assert_eq!(list[1], - 8);
	assert_eq!(list[2], -27);
}


fn test_iterator() {
	let mut list = AvlTreeList::<i32>::new();
	for i in 0 .. 50 {
		list.push(i * i);
	}
	
	let mut iter = list.into_iter();
	for i in 0 .. 50 {
		assert_eq!(*iter.next().unwrap(), i * i);
	}
	assert_eq!(iter.next(), None);
}


// Comprehensively tests all the defined methods.
fn test_against_rust_vec_randomly() {
	let trials = 100_000;
	let mut rng = rand::thread_rng();
	let opcountdist = Range::new(1, 101);
	let valuedist = Range::new(0i32, 1000000i32);
	
	let mut list0 = Vec::<i32>::new();
	let mut list1 = AvlTreeList::<i32>::new();
	let mut size: usize = 0;
	for _ in 0 .. trials {
		let op = Range::new(0, 100).ind_sample(&mut rng);
		
		if op < 1 {  // Clear
			list1.check_structure();
			list0.clear();
			list1.clear();
			size = 0;
			
		} else if op < 2 {  // Set
			if size > 0 {
				let index = Range::new(0, size).ind_sample(&mut rng);
				let val = valuedist.ind_sample(&mut rng);
				list0[index] = val;
				list1[index] = val;
			}
			
		} else if op < 30 {  // Random insertion
			let n = opcountdist.ind_sample(&mut rng);
			for _ in 0 .. n {
				let index = Range::new(0, size + 1).ind_sample(&mut rng);
				let val = valuedist.ind_sample(&mut rng);
				list0.insert(index, val);
				list1.insert(index, val);
			}
			size += n;
			
		} else if op < 50 {  // Ascending insertion
			let n = opcountdist.ind_sample(&mut rng);
			let offset =  Range::new(0, size + 1).ind_sample(&mut rng);
			for i in 0 .. n {
				let val = valuedist.ind_sample(&mut rng);
				list0.insert(offset + i, val);
				list1.insert(offset + i, val);
			}
			size += n;
			
		} else if op < 70 {  // Descending insertion
			let n = opcountdist.ind_sample(&mut rng);
			let offset = Range::new(0, size + 1).ind_sample(&mut rng);
			for _ in 0 .. n {
				let val = valuedist.ind_sample(&mut rng);
				list0.insert(offset, val);
				list1.insert(offset, val);
			}
			size += n;
			
		} else if op < 80 {  // Random deletion
			let n = std::cmp::min(opcountdist.ind_sample(&mut rng), size);
			for _ in 0 .. n {
				let index = Range::new(0, size).ind_sample(&mut rng);
				assert_eq!(list0.remove(index), list1.remove(index));
				size -= 1;
			}
			
		} else if op < 90 {  // Ascending deletion
			if size > 0 {
				let offset = Range::new(0, size).ind_sample(&mut rng);
				let n = std::cmp::min(opcountdist.ind_sample(&mut rng), size - offset);
				for _ in 0 .. n {
					assert_eq!(list0.remove(offset), list1.remove(offset));
				}
				size -= n;
			}
			
		} else if op < 100 {  // Descending deletion
			if size > 0 {
				let offset = Range::new(0, size).ind_sample(&mut rng);
				let n = std::cmp::min(opcountdist.ind_sample(&mut rng), offset + 1);
				for i in 0 .. n {
					assert_eq!(list0.remove(offset - i), list1.remove(offset - i));
				}
				size -= n;
			}
		} else {
			unreachable!();
		}
		
		assert_eq!(list0.len(), size);
		assert_eq!(list1.len(), size);
		if size > 0 {
			let indexdist = Range::new(0, size);
			for _ in 0 .. 10 {
				let index = indexdist.ind_sample(&mut rng);
				assert_eq!(list0[index], list1[index]);
			}
		}
	}
}

File: avl-tree-list/AvlTreeListTest.js
/*
 * AVL tree list test (compiled from TypeScript)
 *
 * Copyright (c) 2022 Project Nayuki. (MIT License)
 * https://www.nayuki.io/page/avl-tree-list
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * - The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 * - The Software is provided "as is", without warranty of any kind, express or
 *   implied, including but not limited to the warranties of merchantability,
 *   fitness for a particular purpose and noninfringement. In no event shall the
 *   authors or copyright holders be liable for any claim, damages or other
 *   liability, whether in an action of contract, tort or otherwise, arising from,
 *   out of or in connection with the Software or the use or other dealings in the
 *   Software.
 */
"use strict";
/*---- Test suite ----*/
const TEST_SUITE_FUNCS = [
    function testPush() {
        let list = new AvlTreeList();
        list.push("January");
        list.push("February");
        list.push("March");
        list.push("April");
        list.push("May");
        list.push("June");
        list.checkStructure();
        assertEquals(6, list.length);
        assertEquals("January", list.get(0));
        assertEquals("February", list.get(1));
        assertEquals("March", list.get(2));
        assertEquals("April", list.get(3));
        assertEquals("May", list.get(4));
        assertEquals("June", list.get(5));
    },
    function testSet() {
        let list = new AvlTreeList();
        for (let i = 0; i < 10; i++)
            list.push(null);
        list.set(0, "zero");
        list.set(1, "ten");
        list.set(2, "twenty");
        list.set(3, "thirty");
        list.set(4, "forty");
        list.set(5, "fifty");
        list.set(6, "sixty");
        list.set(7, "seventy");
        list.set(8, "eighty");
        list.set(9, "ninety");
        assertEquals(10, list.length);
        assertEquals("zero", list.get(0));
        assertEquals("ten", list.get(1));
        assertEquals("twenty", list.get(2));
        assertEquals("thirty", list.get(3));
        assertEquals("forty", list.get(4));
        assertEquals("fifty", list.get(5));
        assertEquals("sixty", list.get(6));
        assertEquals("seventy", list.get(7));
        assertEquals("eighty", list.get(8));
        assertEquals("ninety", list.get(9));
    },
    function testInsertAtBeginning() {
        let list = new AvlTreeList();
        list.insert(0, "Sunday");
        list.insert(0, "Monday");
        list.insert(0, "Tuesday");
        assertEquals(3, list.length);
        assertEquals("Tuesday", list.get(0));
        assertEquals("Monday", list.get(1));
        assertEquals("Sunday", list.get(2));
    },
    function testInsertAtEnd() {
        let list = new AvlTreeList();
        list.insert(0, "Saturday");
        list.insert(1, "Friday");
        list.insert(2, "Thursday");
        list.insert(3, "Wednesday");
        assertEquals(4, list.length);
        assertEquals("Saturday", list.get(0));
        assertEquals("Friday", list.get(1));
        assertEquals("Thursday", list.get(2));
        assertEquals("Wednesday", list.get(3));
    },
    function testInsertAtMiddle() {
        let list = new AvlTreeList();
        list.insert(0, "Up");
        list.insert(1, "Down");
        list.insert(1, "Left");
        list.insert(2, "Right");
        list.insert(1, "Front");
        list.insert(2, "Back");
        assertEquals(6, list.length);
        assertEquals("Up", list.get(0));
        assertEquals("Front", list.get(1));
        assertEquals("Back", list.get(2));
        assertEquals("Left", list.get(3));
        assertEquals("Right", list.get(4));
        assertEquals("Down", list.get(5));
    },
    // Stresses the self-balancing mechanism
    function testInsertManyBeginning() {
        let list = new AvlTreeList();
        for (let i = 99999; i >= 0; i--)
            list.insert(0, i);
        let i = 0;
        for (let iter = list.iterator(); iter.hasNext();) {
            assertEquals(i, iter.next());
            assertEquals(i, list.get(i));
            i++;
        }
    },
    // Stresses the self-balancing mechanism
    function testInsertManyEnd() {
        let list = new AvlTreeList();
        for (let i = 0; i < 100000; i++)
            list.push(i);
        let i = 0;
        for (let iter = list.iterator(); iter.hasNext();) {
            assertEquals(i, iter.next());
            assertEquals(i, list.get(i));
            i++;
        }
    },
    // Adds in a weird binary pattern to stress arrays and linked lists
    function testInsertManyEverywhere() {
        const N = 17;
        let list = new AvlTreeList();
        list.push(0);
        for (let i = N - 1; i >= 0; i--) {
            for (let j = 1 << i, k = 1; j < (1 << N); j += 2 << i, k += 2)
                list.insert(k, j);
        }
        let i = 0;
        for (let iter = list.iterator(); iter.hasNext();) {
            assertEquals(i, iter.next());
            assertEquals(i, list.get(i));
            i++;
        }
    },
    function testNewFromArrayShiftPop() {
        let list = new AvlTreeList([3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5, 8, 9, 7, 9, 3, 2, 3, 8, 4, 6, 2, 6, 4, 3, 3, 8, 3]);
        assertEquals(28, list.length);
        assertEquals(3, list.shift());
        assertEquals(3, list.pop());
        assertEquals(8, list.pop());
        assertEquals(1, list.shift());
        assertEquals(4, list.shift());
        assertEquals(3, list.pop());
        assertEquals(1, list.shift());
        assertEquals(3, list.pop());
        assertEquals(4, list.pop());
        assertEquals(6, list.pop());
        assertEquals(5, list.shift());
        assertEquals(9, list.shift());
        assertEquals(2, list.shift());
    },
    function testRemove() {
        let list = new AvlTreeList();
        let str = "the quick brown fox jumped over the lazy dog";
        for (let i = 0; i < str.length; i++)
            list.push(str.charAt(i));
        assertEquals(str.length, list.length);
        assertEquals("e", list.get(2));
        list.remove(2);
        assertEquals("u", list.get(4));
        list.remove(4);
        assertEquals("q", list.get(3));
        list.remove(3);
        assertEquals(" ", list.get(2));
        list.remove(2);
        assertEquals("f", list.get(12));
        list.remove(12);
        assertEquals(" ", list.get(11));
        list.remove(11);
        assertEquals("n", list.get(10));
        list.remove(10);
        assertEquals("w", list.get(9));
        list.remove(9);
        assertEquals(" ", list.get(11));
        list.remove(11);
        assertEquals("j", list.get(11));
        list.remove(11);
        assertEquals("u", list.get(11));
        list.remove(11);
        assertEquals("x", list.get(10));
        list.remove(10);
        assertEquals("p", list.get(11));
        list.remove(11);
        assertEquals("d", list.get(12));
        list.remove(12);
        assertEquals("e", list.get(11));
        list.remove(11);
        assertEquals("v", list.get(13));
        list.remove(13);
        assertEquals("e", list.get(13));
        list.remove(13);
        assertEquals("l", list.get(19));
        list.remove(19);
        assertEquals("z", list.get(20));
        list.remove(20);
        assertEquals("a", list.get(19));
        list.remove(19);
        assertEquals(" ", list.get(18));
        list.remove(18);
        assertEquals("g", list.get(22));
        list.remove(22);
        str = "thick broom or they do";
        assertEquals(str.length, list.length);
        for (let i = 0; i < str.length; i++)
            assertEquals(str.charAt(i), list.get(i));
        assertEquals("t", list.get(0));
        list.remove(0);
        assertEquals("c", list.get(2));
        list.remove(2);
        assertEquals("k", list.get(2));
        list.remove(2);
        assertEquals(" ", list.get(2));
        list.remove(2);
        assertEquals("b", list.get(2));
        list.remove(2);
        assertEquals("r", list.get(2));
        list.remove(2);
        assertEquals("o", list.get(2));
        list.remove(2);
        assertEquals("o", list.get(2));
        list.remove(2);
        assertEquals("o", list.get(4));
        list.remove(4);
        assertEquals("h", list.get(7));
        list.remove(7);
        assertEquals(" ", list.get(5));
        list.remove(5);
        assertEquals("t", list.get(5));
        list.remove(5);
        assertEquals("o", list.get(9));
        list.remove(9);
        assertEquals(" ", list.get(7));
        list.remove(7);
        assertEquals("y", list.get(6));
        list.remove(6);
        str = "him red";
        assertEquals(str.length, list.length);
        for (let i = 0; i < str.length; i++)
            assertEquals(str.charAt(i), list.get(i));
    },
    function testClear() {
        let list = new AvlTreeList();
        for (let i = 0; i < 20; i++)
            list.push(i * i);
        list.clear();
        assertEquals(0, list.length);
        list.push(-1);
        list.push(-8);
        list.push(-27);
        assertEquals(3, list.length);
        assertEquals(-1, list.get(0));
        assertEquals(-8, list.get(1));
        assertEquals(-27, list.get(2));
    },
    function testSlice() {
        let list = new AvlTreeList([2, 7, 1, 8, 2, 8, 1, 8, 2, 8, 4, 5, 9, 0, 4, 5, 2, 3, 5, 3, 6, 0]);
        let temp = list.slice();
        assertEquals(list.length, temp.length);
        for (let i = 0; i < list.length; i++)
            assertEquals(list.get(i), temp.get(i));
        temp.remove(5);
        temp.set(2, 0);
        assertEquals(list.length - 1, temp.length);
        assertEquals(0, temp.get(2));
        assertEquals(1, list.get(2));
        assertArrayEquals([2, 7, 1, 8], list.slice(0, 4).toArray());
        assertArrayEquals([], list.slice(1, 1).toArray());
        assertArrayEquals([3, 6], list.slice(-3, -1).toArray());
    },
    function testSplice() {
        let list = new AvlTreeList([1, 6, 1, 8, 0, 3, 3, 9, 8, 8, 7, 4, 9, 8, 9, 4, 8, 4, 8, 2, 0, 4, 5, 8, 6, 8, 3, 4, 3]);
        let temp = list.splice(0, 0);
        assertEquals(true, temp instanceof AvlTreeList);
        assertEquals(0, temp.length);
        assertEquals(29, list.length);
        temp = list.splice(-3, 5);
        assertEquals(3, temp.length);
        assertArrayEquals([3, 4, 3], temp.toArray());
        assertEquals(26, list.length);
        temp = list.splice(-27, 2, 5);
        assertArrayEquals([1, 6], temp.toArray());
        assertEquals(25, list.length);
        assertEquals(5, list.get(0));
        temp = list.splice(4, 1, 4, 2, 5);
        assertEquals(5, list.get(0));
        assertEquals(1, list.get(1));
        assertEquals(8, list.get(2));
        assertEquals(0, list.get(3));
        assertEquals(4, list.get(4));
        assertEquals(2, list.get(5));
        assertEquals(5, list.get(6));
        assertEquals(3, list.get(7));
        assertEquals(9, list.get(8));
        assertEquals(8, list.get(9));
    },
    function testForEach() {
        let list = new AvlTreeList([1, 4, 1, 4, 2]);
        let arr = [];
        list.forEach((val, i) => {
            arr.push(i);
            arr.push(val);
        });
        assertArrayEquals([0, 1, 1, 4, 2, 1, 3, 4, 4, 2], arr);
        let obj = {};
        list.forEach(function () { assertEquals(obj, this); }, obj);
    },
    function testIterator() {
        let list = new AvlTreeList();
        for (let i = 0; i < 50; i++)
            list.push(i * i);
        let iter = list.iterator();
        for (let i = 0; i < 50; i++) {
            assertEquals(true, iter.hasNext());
            assertEquals(i * i, iter.next());
        }
        assertEquals(false, iter.hasNext());
    },
    // Tests most of the defined methods.
    function testAgainstJavascriptListRandomly() {
        let list0 = [];
        let list1 = new AvlTreeList();
        let size = 0;
        for (let i = 0; i < 10000; i++) {
            const op = Math.random();
            if (op < 0.01) { // To array and clear
                list1.checkStructure();
                assertArrayEquals(list0, list1.toArray());
                list0 = [];
                list1.clear();
                size = 0;
            }
            else if (op < 0.02) { // Set
                if (size > 0) {
                    const index = Math.floor(Math.random() * size);
                    const val = Math.random();
                    list0[index] = val;
                    list1.set(index, val);
                }
            }
            else if (op < 0.30) { // Random insertion
                const n = Math.floor(Math.random() * 100) + 1;
                for (var j = 0; j < n; j++) {
                    const index = Math.floor(Math.random() * (size + 1));
                    const val = Math.random();
                    list0.splice(index, 0, val);
                    list1.insert(index, val);
                }
                size += n;
            }
            else if (op < 0.50) { // Ascending insertion
                const n = Math.floor(Math.random() * 100) + 1;
                let offset = Math.floor(Math.random() * (size + 1));
                for (let j = 0; j < n; j++, offset++) {
                    const val = Math.random();
                    list0.splice(offset, 0, val);
                    list1.insert(offset, val);
                }
                size += n;
            }
            else if (op < 0.70) { // Descending insertion
                const n = Math.floor(Math.random() * 100) + 1;
                const offset = Math.floor(Math.random() * (size + 1));
                for (let j = 0; j < n; j++) {
                    const val = Math.random();
                    list0.splice(offset, 0, val);
                    list1.insert(offset, val);
                }
                size += n;
            }
            else if (op < 0.80) { // Random deletion
                const n = Math.floor(Math.random() * 100) + 1;
                for (let j = 0; j < n && size > 0; j++, size--) {
                    const index = Math.floor(Math.random() * size);
                    assertEquals(list0[index], list1.get(index));
                    list0.splice(index, 1);
                    list1.remove(index);
                }
            }
            else if (op < 0.90) { // Ascending deletion
                const n = Math.floor(Math.random() * 100) + 1;
                if (size > 0) {
                    const offset = Math.floor(Math.random() * size);
                    for (let j = 0; j < n && offset < size; j++, size--) {
                        assertEquals(list0[offset], list1.get(offset));
                        list0.splice(offset, 1);
                        list1.remove(offset);
                    }
                }
            }
            else if (op < 1.00) { // Descending deletion
                const n = Math.floor(Math.random() * 100) + 1;
                if (size > 0) {
                    let offset = Math.floor(Math.random() * size);
                    for (let j = 0; j < n && offset >= 0; j++, offset--, size--) {
                        assertEquals(list0[offset], list1.get(offset));
                        list0.splice(offset, 1);
                        list1.remove(offset);
                    }
                }
            }
            else
                throw new Error("Assertion error");
            assertEquals(size, list0.length);
            assertEquals(size, list1.length);
            if (size > 0) {
                for (let j = 0; j < 10; j++) {
                    const index = Math.floor(Math.random() * size);
                    assertEquals(list0[index], list1.get(index));
                }
            }
        }
    },
];
/*---- Helper definitions ----*/
function assertEquals(expected, actual) {
    if (expected !== actual)
        throw new Error("Values are unequal");
}
function assertArrayEquals(expected, actual) {
    if (!(expected instanceof Array) || !(actual instanceof Array))
        throw new TypeError("Illegal argument");
    if (expected.length != actual.length)
        throw new Error("Array length mismatch");
    for (let i = 0; i < expected.length; i++)
        assertEquals(expected[i], actual[i]);
}
/*---- Main runner ----*/
(function () {
    let i = 0;
    function iterate() {
        let msg;
        if (i >= TEST_SUITE_FUNCS.length)
            msg = "Finished";
        else {
            msg = TEST_SUITE_FUNCS[i].name + "(): ";
            try {
                TEST_SUITE_FUNCS[i]();
                msg += "Pass";
            }
            catch (e) {
                msg += "Fail - " + e.message;
            }
            i++;
            setTimeout(iterate);
        }
        let li = document.createElement("li");
        li.textContent = msg;
        document.getElementById("results").append(li);
    }
    iterate();
})();
